File: Quill/Classes/AppController.m
//
//  AppController.m
//  Quill
//
//  Created by Shintaro Morikawa on 2013/11/14.
//  Copyright (c) 2013年 Shintaro Morikawa. All rights reserved.
//

#import "AppController.h"
#import "NSPasteboard+NSPasteboard_Stash.h"

@interface AppController () {
    TrieTree *trieTree_;
    KeyMonitor *keyMonitor_;
    __weak NSWindow * _mainWindow;
}
@end

@implementation AppController
- (id)init
{
    self = [super init];
    if (self) {
        trieTree_ = [TrieTree sharedTrieTree];
        keyMonitor_ = [[KeyMonitor alloc] init];
        keyMonitor_.delegte = self;
        _mainWindow = nil;
    }
    return self;
}

- (void)postDeleteAndPasteEvent:(int)nDeletes {
    CGEventSourceRef source = CGEventSourceCreate(kCGEventSourceStateCombinedSessionState);
    
    CGEventRef deleteDown = CGEventCreateKeyboardEvent(source, kVK_Delete, true);
    CGEventRef deleteUp = CGEventCreateKeyboardEvent(source, kVK_Delete, false);
    CGEventRef pasteCommandDown = CGEventCreateKeyboardEvent(source, kVK_ANSI_V, YES);
    CGEventRef pasteCommandUp = CGEventCreateKeyboardEvent(source, kVK_ANSI_V, NO);
    CGEventSetFlags(pasteCommandDown, kCGEventFlagMaskCommand);
    
    CGEventTapLocation loc = kCGSessionEventTap;
    
    for (int i=0; i<nDeletes; i++) {
        CGEventPost(loc, deleteDown);
        CGEventPost(loc, deleteUp);
    }
    CGEventPost(loc, pasteCommandDown);
    CGEventPost(loc, pasteCommandUp);
    
    CFRelease(deleteDown);
    CFRelease(deleteUp);
    CFRelease(pasteCommandUp);
    CFRelease(pasteCommandDown);
    CFRelease(source);
}

#pragma mark
- (void)openMainWindow {
    if (_mainWindow == nil) {
        mainViewController_ = [[MainViewController alloc] initWithNibName:@"MainViewController" bundle:nil];
        _mainWindow = [self createWindowWithContentView:mainViewController_.view];
        [_mainWindow setTitle:@"Quill"];
        mainWindowController_ = [[NSWindowController alloc] initWithWindow:_mainWindow];
    }
    [mainWindowController_ showWindow:self];
    [NSApp activateIgnoringOtherApps:YES];
}

- (NSWindow *)createWindowWithContentView:(NSView *)view {
    NSRect frame = NSMakeRect(0, 0, view.frame.size.width, view.frame.size.height+22);
    NSUInteger styleMask = (NSWindowStyleMaskTitled
                            | NSWindowStyleMaskClosable
                            | NSWindowStyleMaskMiniaturizable
                            /*| NSResizableWindowMask*/);
    NSRect rect = [NSWindow contentRectForFrameRect:frame styleMask:styleMask];
    NSWindow *window = [[NSWindow alloc] initWithContentRect:rect
                                                   styleMask:styleMask
                                                     backing:NSBackingStoreBuffered
                                                       defer:false];
    [window center];
    [window setContentView:view];
    return window;
}

#pragma mark - KeyMonitorDelegate method
- (void)keyUp:(unichar)character {
    if (0<=character && character<256) {
        TrieNode *accept = [trieTree_ stateTransit:character];
        if (accept != NULL) {
            NSString *data = [NSString stringWithUTF8String:accept->data];
            unsigned int depth = accept->depth;

            // Save current pasteboard state
            NSPasteboard *pboard = [NSPasteboard generalPasteboard];
            NSArray *savedPasteboardItems = [pboard save];

            // Paste snippet
            [pboard clearContents];
            [pboard setString:data forType:NSPasteboardTypeString];
            [self postDeleteAndPasteEvent:depth-1];

            [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];

            // Restore current pasteboard state
            [pboard restore:savedPasteboardItems];
        }
    }
}


@end

File: Quill/AppController.swift
/*
 * "Hello Swift, Goodbye Obj-C."
 * Converted by 'objc2swift'
 *
 * https://github.com/yahoojapan/objc2swift
 */

import Foundation
import Cocoa

class AppController: NSObject, KeyMonitorDelegate {
    var mainWindowController: NSWindowController?
    var licenseWindowController: NSWindowController?
    var mainViewController: MainViewController?

    private let trieTree = TrieTreeManager.shared
    private let keyMonitor: KeyMonitor
    private var mainWindow: NSWindow?

    @objc func openMainWindow() {
        licenseWindowController?.close()
        if mainWindow == nil {
            let viewController = MainViewController(nibName: "MainViewController", bundle: nil)
            mainViewController = viewController
            let window = createWindowWithContentView(view: viewController.view)
            window.title = "Quill - \(Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString").flatMap(String.init(describing:)) ?? "")"
            mainWindow = window
            mainWindowController = NSWindowController(window: window)
        }
        mainWindowController?.showWindow(self)
        NSApp.activate(ignoringOtherApps: true)
    }

    @objc func openLicenseWindow() {
        mainWindowController?.close()
        licenseWindowController = LicenseWindowController(windowNibName: "LicenseWindowController")
        licenseWindowController?.showWindow(self)
        NSApp.activate(ignoringOtherApps: true)
    }

    func keyUp(character: Character) {
        let state = trieTree.nextState(key: character)
        if case let .accept(key, value) = state {
            let pboard = NSPasteboard.general
            let savedPasteboardItems = pboard.save()
            pboard.clearContents()
            pboard.setString(value, forType: .string)
            self.postDeleteAndPasteEvent(nDeletes: key.count)
            RunLoop.current.run(until: Date(timeIntervalSinceNow: 0.1))
            pboard.restore(savedPasteboardItems)
        }
    }

    override init() {
        keyMonitor = KeyMonitor()
        super.init()
        keyMonitor.delegte = self
    }

    func postDeleteAndPasteEvent(nDeletes: Int) {
        let source = CGEventSource(stateID: .combinedSessionState)

        let kVK_Delete = 0x33
        let kVK_ANSI_V = 0x09

        guard
            let deleteDown = CGEvent(keyboardEventSource: source, virtualKey: CGKeyCode(kVK_Delete), keyDown: true),
            let deleteUp = CGEvent(keyboardEventSource: source, virtualKey: CGKeyCode(kVK_Delete), keyDown: false),
            let pasteCommandDown = CGEvent(keyboardEventSource: source, virtualKey: CGKeyCode(kVK_ANSI_V), keyDown: true),
            let pasteCommandUp = CGEvent(keyboardEventSource: source, virtualKey: CGKeyCode(kVK_ANSI_V), keyDown: false)
        else { return assertionFailure() }

        pasteCommandDown.flags = CGEventFlags.maskCommand

        (0...nDeletes).forEach { _ in
            deleteDown.post(tap: .cgSessionEventTap)
            deleteUp.post(tap: .cgSessionEventTap)
        }

        pasteCommandDown.post(tap: .cgSessionEventTap)
        pasteCommandUp.post(tap: .cgSessionEventTap)
    }

    func createWindowWithContentView(view: NSView) -> NSWindow {
        let frame = NSMakeRect(0, 0, view.frame.size.width, view.frame.size.height + 22)
        let styleMask: NSWindow.StyleMask = [.titled, .closable, .miniaturizable]
        let rect = NSWindow.contentRect(forFrameRect: frame, styleMask: styleMask)
        let window = NSWindow(contentRect: rect, styleMask: styleMask, backing: .buffered, defer: false)
        window.center()
        window.contentView = view
        return window
    }
}
